Build a regression model.

In [2]:
import pandas as pd
import sqlite3
import statsmodels.api as sm

In [3]:
df = pd.read_csv('../data/merged_data.csv')
df

name  latitude_x  longitude_x  empty_slots  \
0            Queen St E / Woodward Ave   43.665269   -79.319796          5.0   
1          Primrose Ave / Davenport Rd   43.671420   -79.445947         15.0   
2            Queen St. E / Rhodes Ave.   43.666224   -79.317693         10.0   
3                 Bond St / Queen St E   43.653236   -79.376716         22.0   
4             Church St / Alexander St   43.663722   -79.380288          4.0   
..                                 ...         ...          ...          ...   
700                  Grandma Loves You         NaN          NaN          NaN   
701                   Cactus Club Cafe         NaN          NaN          NaN   
702  The Queen And Beaver Public House         NaN          NaN          NaN   
703                        Moon Palace         NaN          NaN          NaN   
704                Liko's Hawaiian BBQ         NaN          NaN          NaN   

     free_bikes  review_count  rating price               location  
0          12.0           NaN     NaN   NaN                    NaN  
1           0.0           NaN     NaN   NaN                    NaN  
2          13.0           NaN     NaN   NaN                    NaN  
3           1.0           NaN     NaN   NaN                    NaN  
4          23.0           NaN     NaN   NaN                    NaN  
..          ...           ...     ...   ...                    ...  
700         NaN          11.0     4.5   NaN  111 Richmond Street W  
701         NaN         601.0     4.0    $$   77 Adelaide Street W  
702         NaN         412.0     4.0    $$          35 Elm Street  
703         NaN          60.0     4.0   NaN     40 Dundas Street W  
704         NaN           4.0     4.0   NaN                    NaN  

[705 rows x 9 columns]

In [16]:
df.dtypes

name             object
latitude_x      float64
longitude_x     float64
empty_slots     float64
free_bikes      float64
review_count    float64
rating          float64
price            object
location         object
dtype: object

Provide model output and an interpretation of the results. 

In [17]:
from sklearn.preprocessing import LabelEncoder
cols = ['free_bikes', 'review_count']
le = LabelEncoder()
for col in cols:
    df[col] = le.fit_transform(df[col])

In [18]:
X = df['free_bikes']
y = df['review_count']

In [19]:
X = sm.add_constant(X)
model = sm.OLS(y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           review_count   R-squared:                       0.190
Model:                            OLS   Adj. R-squared:                  0.188
Method:                 Least Squares   F-statistic:                     164.5
Date:                Wed, 16 Aug 2023   Prob (F-statistic):           5.57e-34
Time:                        20:41:37   Log-Likelihood:                -1408.5
No. Observations:                 705   AIC:                             2821.
Df Residuals:                     703   BIC:                             2830.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         19.7301      0.105    188.673      0.000      19.525      19.935
free_bikes    -0.1106      0.009    -12.824      0.000      -0.128      -0.094
==============================================================================
Omnibus:                      821.938   Durbin-Watson:                   0.786
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            58055.739
Skew:                          -5.775   Prob(JB):                         0.00
Kurtosis:                      45.930   Cond. No.                         18.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

From the P value of 0.000 and R-squared of 0.190 - the model is not a good fit

In [20]:
from sklearn.preprocessing import LabelEncoder
cols = ['empty_slots', 'rating']
le = LabelEncoder()
for col in cols:
    df[col] = le.fit_transform(df[col])

In [21]:
X = df['empty_slots']
y = df['rating']

In [22]:
X = sm.add_constant(X)
model = sm.OLS(y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 rating   R-squared:                       0.276
Model:                            OLS   Adj. R-squared:                  0.275
Method:                 Least Squares   F-statistic:                     268.5
Date:                Wed, 16 Aug 2023   Prob (F-statistic):           2.39e-51
Time:                        20:45:33   Log-Likelihood:                -286.16
No. Observations:                 705   AIC:                             576.3
Df Residuals:                     703   BIC:                             585.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const           3.1950      0.021    150.467      0.000       3.153       3.237
empty_slots    -0.0264      0.002    -16.385      0.000      -0.030      -0.023
==============================================================================
Omnibus:                      597.520   Durbin-Watson:                   0.628
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            14093.441
Skew:                          -3.766   Prob(JB):                         0.00
Kurtosis:                      23.568   Cond. No.                         20.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

From the P value of 0.000 and R-squared of 0.276 - the model is not a good fit

In [23]:
from sklearn.preprocessing import LabelEncoder
cols = ['empty_slots', 'free_bikes']
le = LabelEncoder()
for col in cols:
    df[col] = le.fit_transform(df[col])

In [24]:
X = df['empty_slots']
y = df['free_bikes']

In [25]:
X = sm.add_constant(X)
model = sm.OLS(y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             free_bikes   R-squared:                       0.039
Model:                            OLS   Adj. R-squared:                  0.037
Method:                 Least Squares   F-statistic:                     28.30
Date:                Wed, 16 Aug 2023   Prob (F-statistic):           1.40e-07
Time:                        20:48:20   Log-Likelihood:                -2435.0
No. Observations:                 705   AIC:                             4874.
Df Residuals:                     703   BIC:                             4883.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
===============================================================================
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
const          11.1027      0.447     24.814      0.000      10.224      11.981
empty_slots    -0.1804      0.034     -5.320      0.000      -0.247      -0.114
==============================================================================
Omnibus:                      215.263   Durbin-Watson:                   1.160
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              554.694
Skew:                           1.568   Prob(JB):                    3.55e-121
Kurtosis:                       6.007   Cond. No.                         20.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

The R-squared value of 0.039 and pP-value of 0.000 shows that model is not a good fit

# Stretch

How can you turn the regression model into a classification model?